src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -1042,21 +1046,46 @@ List *
 SequenceDDLCommandsForTable(Oid relationId)
 {
 	List *sequenceDDLList = NIL;
-	List *ownedSequences = GetSequencesOwnedByRelation(relationId);
+
+	List *attnumList = NIL;
+	List *dependentSequenceList = NIL;
+	GetDependentSequencesWithRelation(relationId, &attnumList, &dependentSequenceList, 0);
+
 	char *ownerName = TableOwner(relationId);
 
-	Oid sequenceOid = InvalidOid;
-	foreach_oid(sequenceOid, ownedSequences)
+	ListCell *attnumCell = NULL;
+	ListCell *dependentSequenceCell = NULL;
+	forboth(attnumCell, attnumList, dependentSequenceCell, dependentSequenceList)
 	{
+		AttrNumber attnum = lfirst_int(attnumCell);
+		Oid sequenceOid = lfirst_oid(dependentSequenceCell);
+
 		char *sequenceDef = pg_get_sequencedef_string(sequenceOid);
 		char *escapedSequenceDef = quote_literal_cstr(sequenceDef);
 		StringInfo wrappedSequenceDef = makeStringInfo();
 		StringInfo sequenceGrantStmt = makeStringInfo();
 		char *sequenceName = generate_qualified_relation_name(sequenceOid);
 		Form_pg_sequence sequenceData = pg_get_sequencedef(sequenceOid);
-		Oid sequenceTypeOid = sequenceData->seqtypid;
+		Oid sequenceTypeOid = GetAttributeTypeOid(relationId, attnum);
 		char *typeName = format_type_be(sequenceTypeOid);
 
+		/* get sequence address */
+		ObjectAddress sequenceAddress = { 0 };
+		ObjectAddressSet(sequenceAddress, RelationRelationId, sequenceOid);
+		EnsureDependenciesExistOnAllNodes(&sequenceAddress);
+
+		/* alter the sequence's data type in the coordinator if needed */
+		Oid currentSequenceTypeOid = sequenceData->seqtypid;
+		if (currentSequenceTypeOid != sequenceTypeOid)
+		{
+			UnmarkObjectDistributed(&sequenceAddress);
+			StringInfo sequenceAlterTypeStmt = makeStringInfo();
+			appendStringInfo(sequenceAlterTypeStmt, "ALTER SEQUENCE %s AS %s",
+							 sequenceName,
+							 typeName);
+			ExecuteQueryViaSPI(sequenceAlterTypeStmt->data, SPI_OK_UTILITY);
```

Comment:
Would be good to avoid the Citus hooks here and use something like:
```c
AlterSeqStmt *alterSequenceStatement = makeNode(AlterSeqStmt);
alterSequenceStatement->sequence = makeRangeVar(schemaName, sequenceName, -1);
Node *asTypeNode = (Node *) makeTypeNameFromOid(currentSequenceTypeOid, -1);
SetDefElemArg(alterSequenceStatement, "as", asTypeNode);
```
I'm also kind of wondering whether we should be doing this as soon as we start using a sequence in a distributed table (e.g. before we do the type sanity checks), but it's not a blocker.

Reply:
Thanks!